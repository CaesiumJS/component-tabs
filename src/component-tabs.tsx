import React, {useState} from 'react'

interface TabsProps{
  // The tab to start on, defaults to 0.
  current?: number

  /* A Function called when ever the tabe is changed. Useful if you want to set url fragments etc... */
  onChange?: (newTab: number) => void


  /* Class names to use in the components */
  classNames?: {
    container?: string
    tabList?: string
    tabListEntry?: string
    activeTabListEntry?: string
    currentTab?: string
  }

  children: React.ReactElement<TabProps>[]
}

interface TabProps{
  title: string
}

export const Tabs: React.FC<TabsProps> = ({current, children, onChange, classNames}) => {
  const [currentTab, setTab] = useState(current ? current : 0)

  return <div className={classNames && classNames.container ? classNames.container : 'caesium-tabs-container'}>
    <ul className={classNames && classNames.tabList ? classNames.tabList : 'caesium-tabs-tab-list'}>
      {children.map((tab, i) => {
        let className = (classNames && classNames.tabListEntry ? classNames.tabListEntry : 'caesium-tabs-tab-list-entry')

        if(i === currentTab){
          className = (classNames && classNames.activeTabListEntry ? classNames.activeTabListEntry : 'caesium-tabs-tab-list-active-entry')
        }

        return <li key={i} className={className} onClick={() => {
          setTab(i)
          if(onChange) onChange(i)
        }}>{tab.props.title}</li>
      })}
    </ul>
    <div className={classNames && classNames.currentTab ? classNames.currentTab : 'caesium-tabs-currentTab'}>
      {children[currentTab]}
    </div>
  </div>
}

export const Tab: React.FC<TabProps> = ({title, children}) => {
  return <>{children}</>
}