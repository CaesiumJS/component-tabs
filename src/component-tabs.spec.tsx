import React from 'react'
import {render, cleanup, fireEvent} from '@testing-library/react'

import {Tabs, Tab} from './component-tabs'

describe('Component Tabs', () => {
  afterEach(() => {
    cleanup()
  })

  it('should only place the rendered tab into the DOM', () => {
    let tab = 0

    const TestComponent: React.FC = () => {
      return <Tabs onChange={(newTab) => {
        tab = newTab
      }}>
        <Tab title="Tab 1">
          <p>FOO</p>
        </Tab>
        <Tab title="Tab 2">
          <p>Bar</p>
        </Tab>
      </Tabs>
    }

    const {container, getByText} = render(<TestComponent />)

    expect(container).toMatchSnapshot()
    expect(getByText('FOO')).toBeInTheDocument()

    fireEvent.click(getByText('Tab 2'))

    expect(getByText('Bar')).toBeInTheDocument()
    expect(tab).toBe(1)
  })

  it('should allow you to set classes for components', () => {
    const TestComponent: React.FC = () => {
      return <Tabs classNames={{
        container: 'test1',
        tabList: 'test2',
        tabListEntry: 'test3',
        activeTabListEntry: 'test4',
        currentTab: 'test5'
      }}>
        <Tab title="Tab 1">
          <p>FOO</p>
        </Tab>
        <Tab title="Tab 2">
          <p>Bar</p>
        </Tab>
      </Tabs>
    }

    const {container} = render(<TestComponent />)

    expect(container).toMatchSnapshot()
  })
})