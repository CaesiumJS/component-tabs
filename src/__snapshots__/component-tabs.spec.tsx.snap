// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Tabs should allow you to set classes for components 1`] = `
<div>
  <div
    class="test1"
  >
    <ul
      class="test2"
    >
      <li
        class="test4"
      >
        Tab 1
      </li>
      <li
        class="test3"
      >
        Tab 2
      </li>
    </ul>
    <div
      class="test5"
    >
      <p>
        FOO
      </p>
    </div>
  </div>
</div>
`;

exports[`Component Tabs should only place the rendered tab into the DOM 1`] = `
<div>
  <div
    class="caesium-tabs-container"
  >
    <ul
      class="caesium-tabs-tab-list"
    >
      <li
        class="caesium-tabs-tab-list-active-entry"
      >
        Tab 1
      </li>
      <li
        class="caesium-tabs-tab-list-entry"
      >
        Tab 2
      </li>
    </ul>
    <div
      class="caesium-tabs-currentTab"
    >
      <p>
        FOO
      </p>
    </div>
  </div>
</div>
`;
